stix_shifter_utils/stix_translation/src/json_to_stix/json_to_stix_translator.py

```diff
@@ -190,169 +82,225 @@ def gen_dict_extract(self, key, var):
-                                "cust_object"}
-                    if to_map is None or to_map == '':
-                        self.logger.debug("Removing invalid value '{}' for {}".format(to_map, ds_key))
-                        return
-                    self._handle_cybox_key_def(cust_obj["key"], observation, to_map, object_map, cust_obj["object"])
-            else:
-                self.logger.debug('{} is not found in map, skipping'.format(ds_key))
-            return
-
-        if isinstance(to_map, dict):
-            self.logger.debug('{} is complex; descending'.format(to_map))
-            # If the object is complex we must descend into the map on both sides
-            for key in to_map.keys():
-                self._transform(object_map, observation, ds_map[ds_key], key, to_map)
-            return
-
-        # if the datasource fields is a collection of json object than we need to unwrap it and create multiple objects
-        if isinstance(to_map, list):
-            self.logger.debug('{} is a list; unwrapping.'.format(to_map))
-            for item in to_map:
-                if isinstance(item, dict):
-                    new_ds_map = self._update_object_key(ds_map[ds_key], to_map.index(item))
-                    for field in item.keys():
-                        self._transform(object_map, observation, new_ds_map, field, item)
-        
-        generic_hash_key = ''
-
-        # get the stix keys that are mapped
-        ds_key_def_obj = ds_map[ds_key]
-        if isinstance(ds_key_def_obj, list):
-            ds_key_def_list = ds_key_def_obj
-        else:
-            # Use callback function to run module-specific logic to handle unknown filehash types
-            if self.callback:
-                try:
-                    generic_hash_key = self.callback(obj, ds_key, ds_key_def_obj['key'], self.options)
-                except(Exception):
-                    return
-
-            ds_key_def_list = [ds_key_def_obj]
 
 
-        for ds_key_def in ds_key_def_list:
-            if ds_key_def is None or 'key' not in ds_key_def:
-                self.logger.debug('{} is not valid (None, or missing key)'.format(ds_key_def))
-                continue
+    def _valid_stix_value(self, observable_key, stix_value):
+        """
+        Checks that the given STIX value is valid for this STIX property
+        :param observable_key: the STIX property name
+        :param stix_value: the STIX value translated from the input object
+        :return: whether STIX value is valid for this STIX property
+        :rtype: bool
+        """
+        try: 
+            if observable_key in self.properties and 'valid_regex' in self.properties[observable_key]:
+                pattern = re.compile(self.properties[observable_key]['valid_regex'])
+                match = pattern.match
+                if not match(str(stix_value)):
+                    return False
+            return True
+        except Exception as e:
```

Comment:
its better to print some debug log when this Exception hits

Reply:
then how would we know if this validity check fails. is raising exception less costly?

Reply:
as we discussed, I will add the debug because it is necessary.